### DISCUSSION TITLE
6주차 Discussion

### DISCUSSION THREAD

**Discussion by chloe-codes1** [slippStudy/java-network-study#12] [General]
## 📢 책을 읽으면서 궁금한 점들을 자유롭게 적고 대답해 주세요

6주차 진행 방식

> 진행자: 윤혁님, 주현님
> 

- 11 ~ 12장 읽기
- Discussion 에 질문 or 인상 깊은 점 2개 적어오기

**Comment by jinwookh** [comment]
1. Multicast는 내부 통신용인걸까?
멀티캐스트의 목적지 그룹은 244.0.0.0 ~ 239.255.255.255에 속한다고 한다.
그렇다면 예로 들어 12.7.7.7, 28.12.2.3, 91.129.2.2 이 세 개의 호스트가 서로 multicast로 소통하려면 어떻게 개발해야 할까?

2. 319p의 예제 12-1 NonBuffer, 12-2 SmallBuffer, 12-3 FullBuffer는 버퍼가 없으면 성능이 떨어진다는 것을 보여주기 위한 예시다.
그런데 12-1 NonBuffer도 read할 때 커널의 버퍼를 사용하기 때문에 버퍼를 사용하는 예시 아닌가?

3. mmap을 사용하면 read, write를 호출할 필요가 없기 때문에 성능상 이점이 있는 것으로 책에 나와 있다.(327p)
mmap, read 둘 다 메모리를 읽는다는 점은 똑같은데 mmap이 정말로 성능상 이점이 있을까?

**Reply by jinwookh** [reply]
mmap vs read
https://stackoverflow.com/questions/5588605/mmap-vs-read

**Comment by LenKIM** [comment]
1. Files.copy 의 동작원리는 어떻게 될까? 아주 살짝 봤는데, Buffer 을 사용하지 않는다.
2. 시스템 콜을 줄이기 위해 Java NIO  이야기를 다시해보자. 더 성능을 높일 수 있는  방법이 있을까?

3. 운영체제 공부할 때도 그렇고, 지금도 그렇고... 가상 메모리라는게 무슨 말이지? 가상으로 띄어놓은 메모리. 이건 물리 메모리가 아니라, 물리 메모리의 주소를 가진 파일 인건가?

**Comment by yangyunhyeuk** [comment]
1. 멀티캐스팅을 위한 서버는 기존의 UDP 방식의 서버 프로그래밍과 거의 같으며 클라이언트의 IP를 DatagramPacket에 지정하는 것이 아니라 멀티캐스트 그룹에 해당하는 IP를 지정하는 부분만 다르며, 멀티캐스팅을 위한 클라이언트 역시 기존의 UDP 방식의 클라이언트 프로그래밍과 거의 같으며 다른 점이 있다면 DatagramSocket이 아니라 MulticastSocket을 사용한다는 것과 멀티캐스트 그룹에 참여하기 위한 코드가 추가되는 부분이 신기했다.  

2. 자바 IO가 느리다고 얘기되는 이유는 무엇일까?

**Comment by kangbada0728** [comment]
1. 멀티캐스팅을 위한 서버가 기존의 UDP 서버 프로그래밍과 거의 같다고 하였는데, 실제 서비스들에서도 이렇게 만들고 있을까? 유튜브에서 저장된 영상을 볼때는 TCP 로 연결을 하는 것을 학부시절에 확인했었다.

2. 공유자원을 다루기 위해 많은 애를 쓰는것을 보며 함수형 프로그래밍의 장점이 느껴진다.....

**Comment by chloe-codes1** [comment]
### 궁금한 내용

- 현대 OS에서 가상 메모리를 효과적으로 관리하기 위해 개선된 방법론이 있을까?

### 인상깊은 점

- java.nio가 어떤 요구사항에 의해 등장했고, nio의 근간이 되는 개념들 (Buffer, Channel, Selector) 에 대해 알아볼 수 있어 좋았다
